### PR TITLE
update library to use span without allocating a new array

### PR DESCRIPTION
**Pull request by tanishqjasoria**

### CODE REVIEW

File: src/Nethermind/Nethermind.Trie.Test/RustVerkleLibTest.cs [hunk 1]
```diff
@@ -75,21 +76,60 @@ public class RustVerkleLibTest
...
+        {
+            Span<byte> one = new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1};
+            Span<byte> one32 = new byte[]{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
+
+            IntPtr trie = RustVerkleLib.VerkleTrieNew();
+
+            RustVerkleLib.VerkleTrieInsert(trie, one, one32);
+            RustVerkleLib.VerkleTrieInsert(trie, one32, one);
+
+            RustVerkleLib.VerkleTrieGetSpan(trie, one32).ToArray().Should().BeEquivalentTo(one.ToArray());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie.Test/RustVerkleLibTest.cs:110]: Lets add extension methods like: 

public static GenericCollectionAssertions<T> Should<T>(this ReadOnlySpan<T> value)
    => value.ToArray().Should();

public static StringAssertions Should<T>(this Span<char> value)
    => value.ToString().Should();

And same for BeEquivalentTo

File: src/Nethermind/Nethermind.Trie/RustVerkleLib.cs [hunk 2]
```diff
@@ -16,136 +16,219 @@
...
+        int valueLength = value.Length;
+        if (valueLength != 32)
         {
-            return verkle_trie_new();
+            throw new InvalidOperationException("Value length must be less than 32");
         }
-        
-        public static void VerkleTrieInsert(IntPtr verkleTrie, byte[] key, byte[] value)
+
+        fixed (byte* pKey = &MemoryMarshal.GetReference(key))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/RustVerkleLib.cs:77]: If key and value are Span from stackalloc then we don't need fixing it, they a re on stack already fixed. If they are on heap (slice of an array) then we need fixing them. Maybe we should check if there is performance penalty here and differentiate between those scenarios?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Trie/RustVerkleLib.cs (hunk 2 above)
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Trie/RustVerkleLib.cs:77]: yes, this makes sense. I will check for the performance penalties and update this

Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Trie/RustVerkleLib.cs:77]: I am not sure how to check if the Span is on stack?
also the function signature have the data type as Span<byte>, so in the function I am not able to convert he Span<byte> to byte * without using fixed block

**Comment by LukaszRozmej** [comment]
Still some performance can probably be gained

**Pull request merged**